### PR TITLE
Slack - handle rate limit

### PR DESCRIPTION
**Pull request by liorblob**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/19558
fixes: https://github.com/demisto/etc/issues/19819
fixes: https://github.com/demisto/etc/issues/19355
https://github.com/demisto/etc/issues/19947

## Description
Handling rate limit errors in Slack while allowing to specify:
* Maximum time to wait
* paginated objects buffer

## Does it break backward compatibility?
   - No

## Must have
- [X] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Technical writer review
Mention and link to the files that require a technical writer review.
`Slack.yml`
`CHANGELOG.md`
`README.md`

### CODE REVIEW

File: Integrations/Slack/Slack.py [hunk 1]
```diff
@@ -187,6 +197,54 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
     demisto.setIntegrationContext(integration_context)
 
 
+def send_slack_request_sync(client: slack.WebClient, method: str, http_verb: str = 'POST', file: dict = None,
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: please add docstring

File: Integrations/Slack/Slack.py [hunk 2]
```diff
@@ -187,6 +197,54 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
...
+                            body: dict = None) -> SlackResponse:
+    while True:
+        try:
+            if http_verb == 'POST':
+                if file:
+                    response = client.api_call(method, files={"file": file}, data=body)
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: why for post we use `data` and for other we use `json`?

File: Integrations/Slack/Slack.py [hunk 3]
```diff
@@ -187,6 +197,54 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
     demisto.setIntegrationContext(integration_context)
 
 
+def send_slack_request_sync(client: slack.WebClient, method: str, http_verb: str = 'POST', file: dict = None,
+                            body: dict = None) -> SlackResponse:
+    while True:
+        try:
+            if http_verb == 'POST':
+                if file:
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: dont use reserved words, use `file_` instead

File: Integrations/Slack/Slack.py [hunk 4]
```diff
@@ -187,6 +197,54 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
...
+                if retry_after < MAX_LIMIT_TIME:
+                    time.sleep(retry_after)
+                    continue
+            raise
+        break
+
+    return response
+
+
+async def send_slack_request_async(client: slack.WebClient, method: str, http_verb: str = 'POST', file: dict = None,
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: lets discuss

File: Integrations/Slack/Slack.py [hunk 5]
```diff
@@ -660,16 +749,21 @@ def extract_entitlement(entitlement: str, text: str) -> Tuple[str, str, str, str
 
     if not data:
         data = 'Sorry, I could not perform the selected operation.'
-    im = await client.im_open(user=user.get('id'))
+    body = {
+        'user': user.get('id')
+    }
+    im = (await send_slack_request_async(client, 'im.open', body=body))
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: whats the parentheses here for?

File: Integrations/Slack/Slack_test.py [hunk 6]
```diff
@@ -334,80 +334,102 @@ def setup():
...
+
+
+@pytest.mark.asyncio
+async def test_get_slack_name_channel(mocker):
+    from Slack import get_slack_name
+
+    # Set
+
+    @asyncio.coroutine
+    def api_call(method: str, http_verb: str = 'POST', file: dict = None, params=None, json=None, data=None):
```
Comment (Itay4) [inline Integrations/Slack/Slack_test.py]: cant we define this function once and use it in all tests? or you want as many code lines as possible?

File: Integrations/Slack/Slack.py (hunk 2 above)
Comment (liorblob) [reply Integrations/Slack/Slack.py]: That's how they send files

File: Integrations/Slack/Slack_test.py (hunk 6 above)
Comment (liorblob) [reply Integrations/Slack/Slack_test.py]: It differs between some tests... and Yes.

File: Integrations/Slack/Slack.py [hunk 7]
```diff
@@ -187,6 +203,85 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
...
+
+def send_slack_request_sync(client: slack.WebClient, method: str, http_verb: str = 'POST', file_: dict = None,
+                            body: dict = None) -> SlackResponse:
+    """
+    Sends a request to slack API while handling rate limit errors.
+    :param client: The slack client.
+    :param method: The method to use.
+    :param http_verb: The HTTP method to use.
+    :param file_: A file to send.
+    :param body: The request body.
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: add return value

File: Integrations/Slack/Slack.py [hunk 8]
```diff
@@ -187,6 +203,85 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
...
+
+async def send_slack_request_async(client: slack.WebClient, method: str, http_verb: str = 'POST', file_: dict = None,
+                                   body: dict = None) -> SlackResponse:
+    """
+    Sends an async request to slack API while handling rate limit errors.
+    :param client: The slack client.
+    :param method: The method to use.
+    :param http_verb: The HTTP method to use.
+    :param file_: A file to send.
+    :param body: The request body.
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: add return value

File: Integrations/Slack/Slack.py [hunk 9]
```diff
@@ -1644,19 +1649,27 @@ def init_globals():
...
-    global BOT_NAME, BOT_ICON_URL, MAX_LIMIT_TIME, PAGINATED_COUNT
+    global BOT_NAME, BOT_ICON_URL, MAX_LIMIT_TIME, PAGINATED_COUNT, SSL_CONTEXT
+
+    VERIFY_CERT = not demisto.params().get('unsecure', False)
+    if not VERIFY_CERT:
+        SSL_CONTEXT = ssl.create_default_context()
+        SSL_CONTEXT.check_hostname = False
+        SSL_CONTEXT.verify_mode = ssl.CERT_NONE
+    else:
+        SSL_CONTEXT = None  # type: ignore[assignment]
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: can you explain why this ?

File: Integrations/Slack/CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,5 @@
 ## [Unreleased]
-
+Added rate limit handling
```
Comment (glicht) [inline Integrations/Slack/CHANGELOG.md]:
```suggestion
Added rate limit handling.
```

File: Integrations/Slack/Slack.py [hunk 11]
```diff
@@ -187,6 +205,87 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
...
+                    response = client.api_call(method, files={"file": file_}, data=body)
+                else:
+                    response = client.api_call(method, json=body)
+            else:
+                response = client.api_call(method, http_verb='GET', params=body)
+        except SlackApiError as api_error:
+            response = api_error.response
+            if 'Retry-After' in response.headers:
+                retry_after = int(response.headers['Retry-After'])
+                if retry_after < MAX_LIMIT_TIME:
```
Comment (glicht) [inline Integrations/Slack/Slack.py]: Logic seems flawed here. You should have a counter: `total_retry_time` which increases with each try. So if we did a request and got back retry after 20 seconds and then we do another request and get back retry after 45 seconds  we won't try the second request if max limit is 60.

File: Integrations/Slack/Slack.py [hunk 12]
```diff
@@ -187,6 +205,87 @@ def set_to_latest_integration_context(key: str, value, wait: bool = False):
...
+                    response = await client.api_call(method, files={"file": file_}, data=body)
+                else:
+                    response = await client.api_call(method, json=body)
+            else:
+                response = await client.api_call(method, http_verb='GET', params=body)
+        except SlackApiError as api_error:
+            response = api_error.response
+            if 'Retry-After' in response.headers:
+                retry_after = response.headers['Retry-After']
+                if retry_after < MAX_LIMIT_TIME:
```
Comment (glicht) [inline Integrations/Slack/Slack.py]: Same comment about `total_retry_time`.

File: Integrations/Slack/CHANGELOG.md (hunk 10 above)
Comment (liorblob) [reply Integrations/Slack/CHANGELOG.md]: Gonna leave it to @kirbles19

File: Integrations/Slack/Slack.py [hunk 13]
```diff
@@ -480,16 +609,16 @@ def check_for_answers(now: datetime):
             'token': BOT_TOKEN,
             'entitlement': question.get('entitlement')
         }
-        res = requests.post(ENDPOINT_URL, data=json.dumps(body), headers=headers, verify=VERIFY_CERT)
+        res = requests.post(ENDPOINT_URL, data=json.dumps(body), headers=headers, proxies=PROXIES, verify=VERIFY_CERT)
```
Comment (Itay4) [inline Integrations/Slack/Slack.py]: can we test that we send proxies here?

**Comment by liorblob** [comment]
@kirbles19 any progress on this?

**Comment by Itay4** [comment]
@richardbluestone please review

**Comment by richardbluestone** [comment]
Files reviewed.

File: Integrations/Slack/CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,5 @@
 ## [Unreleased]
-
```
Comment (Itay4) [inline Integrations/Slack/CHANGELOG.md]: add to changelog the proxy change

**Comment by liorblob** [comment]
@richardbluestone We had to add a parameter called Proxy URL. Could you review it in the Slack YAML file and chanelog please?

**Pull request merged**